iina/FFmpegController.mm

```diff
@@ -0,0 +1,220 @@
+//
+//  FFmpegController.mm
+//  iina
+//
+//  Created by Saagar Jha on 2/16/19.
+//  Copyright © 2019 lhc. All rights reserved.
+//
+
+#import "FFmpegController.h"
+
+// Perhaps this will work someday? Until then, we need to keep around our dummy
+// interface in the header file.
+//#import "IINA-Swift.h"
+
+#import <algorithm>
+#import <memory>
+
+extern "C" {
+#import <libavcodec/avcodec.h>
+#import <libavformat/avformat.h>
+#import <libavutil/avutil.h>
+#import <libavutil/imgutils.h>
+#import <libswscale/swscale.h>
+}
+
+#define DEFAULT_THUMBNAIL_COUNT 100
+#define DEFAULT_THUMBNAIL_WIDTH 240.0
+
+struct AVFormatContextCloser {
+  void operator()(AVFormatContext *formatContext) {
+    avformat_close_input(&formatContext);
+  }
+};
+
+struct AVCodecContextDeleter {
+  virtual void operator()(AVCodecContext *codecContext) {
+    avcodec_free_context(&codecContext);
+  }
+};
+
+struct AVCodecContextDeleterCloser : AVCodecContextDeleter {
+  void operator()(AVCodecContext *codecContext) override {
+    AVCodecContextDeleter()(codecContext);
+    avcodec_close(codecContext);
+  }
+};
+
+struct AVFrameDeleter {
+  void operator()(AVFrame *frame) {
+    av_frame_free(&frame);
+  }
+};
+
+struct AVDeleter {
+  void operator()(uint8_t *pointer) {
+    av_free(pointer);
+  }
+};
+
+struct SwsContextDeleter {
+  void operator()(SwsContext *swsContext) {
+    sws_freeContext(swsContext);
+  }
+};
+
+struct AVPacketUnrefer {
+  void operator()(AVPacket *packet) {
+    av_packet_unref(packet);
+  }
+};
+
+@implementation FFmpegController (Bridge)
+
+- (BOOL)synchronouslyGenerateThumbnailsForFileAtPath:(NSString *)path {
+  std::unique_ptr<AVFormatContext, AVFormatContextCloser> formatContext;
+  auto raw_formatContext = formatContext.release();
+  if (avformat_open_input(&raw_formatContext, path.fileSystemRepresentation, nullptr, nullptr) < 0) {
+    return NO;
+  }
+  formatContext.reset(raw_formatContext);
+
+  if (avformat_find_stream_info(formatContext.get(), nullptr) < 0) {
+    return NO;
+  }
+  auto videoStreamIndex = std::find_if(formatContext->streams, formatContext->streams + formatContext->nb_streams, [](const AVStream *stream) {
+    return stream->codecpar->codec_type == AVMEDIA_TYPE_VIDEO;
+  });
+  if (videoStreamIndex - formatContext->streams < formatContext->nb_streams) {
```

Comment:
This condition seems to always trigger so it exits here. Though videoStreamIndex looks to be the correct `AVStream**`.

Reply:
I've changed the condition to `videoStreamIndex < formatContext->streams + formatContext->nb_streams`.

Reply:
Looking again, since `find_if` returns `last` if it doesn't find anything I believe it should actually be `videoStreamIndex == formatContext->streams + formatContext->nb_streams` to catch the case of it not finding. What's there right now still exits early (and my bad in comment above).

Reply:
The correct condition is `videoStreamIndex >= formatContext->streams + formatContext->nb_streams`.